File: java/0662_widthOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        var record = new ArrayList<Integer>();
        int[] res = {1};
        traverse(root,record,res,0,0);
        return res[0];
    }
    public void traverse(TreeNode root, List<Integer> record, int[] res, int depth, int position)
    {
        if(root==null)
            return;
        if(record.size()<depth+1)
            record.add(position);
        res[0] = Math.max(res[0],position-record.get(depth)+1);
        traverse(root.left,record,res,depth+1,position*2);
        traverse(root.right,record,res,depth+1,position*2+1);
    }
}

File: cpp/0662_widthOfBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// DFS，每一层遍历一个节点就更新一次宽度，最后在层-宽度数组里取最大值
// ....必须用unsigned long long不然会溢出。。
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        vector<long> record; // (depth-first_position) 
        unsigned long long res=0;
        traverse(root,record,res,0,0);
        return (int)res;
    }
    void traverse(TreeNode* root, vector<long> &record, unsigned long long &res, int depth, unsigned long long position)
    {
        if(root==nullptr)
            return;
        // record the first element in this layer
        if(record.size()<depth+1)
            record.push_back(position);
        
        // update max depth with position
        res = max(res,position-record[depth]+1);
        traverse(root->left,record,res,depth+1,position*2);
        traverse(root->right,record,res,depth+1,position*2+1);
    }
};
